File: scripts/test-launcher.ts
import { mkdir, mkdtemp, rm } from "node:fs/promises";
import { homedir } from "node:os";
import { resolve } from "node:path";

// Exercise the real launcher/bridge/frontend on a private bus, with a local model
// fixture. Its lifetime is external to the launcher and must survive shutdown.
const project = resolve(import.meta.dir, "..");
await mkdir(`${project}/.cache/rime-profiles`, { recursive: true });
const profile = await mkdtemp(`${project}/.cache/rime-profiles/launcher-`);
let requests = 0;
const model = Bun.serve({
  hostname: "127.0.0.1",
  port: 0,
  async fetch(request) {
    const path = new URL(request.url).pathname;
    if (path === "/health") return Response.json({ status: "ok" });
    if (path === "/v1/models")
      return Response.json({ data: [{ id: "qwen3-1.7b" }] });
    if (path !== "/v1/chat/completions")
      return new Response(null, { status: 404 });
    requests++;
    const body = (await request.json()) as { messages: { content: string }[] };
    const line = body.messages[1]?.content
      .split("\n")
      .find((line) => line.includes('"程式"'));
    await Bun.sleep(100);
    return Response.json({
      choices: [
        {
          message: {
            content: JSON.stringify({ best: Number.parseInt(line ?? "1", 10) }),
          },
        },
      ],
    });
  },
});
try {
  const child = Bun.spawn(
    [
      "dbus-run-session",
      "--",
      "python",
      "tests/ibus_integration.py",
      profile,
      "--launcher",
    ],
    {
      cwd: project,
      env: {
        ...process.env,
        RIWEN_MODEL_URL: `http://127.0.0.1:${model.port}/v1/chat/completions`,
        RIWEN_PROFILE: profile,
        RIWEN_RIME_SOURCE: resolve(
          process.env.XDG_CONFIG_HOME ?? `${homedir()}/.config`,
          "ibus/rime",
        ),
      },
      stdout: "inherit",
      stderr: "inherit",
    },
  );
  if ((await child.exited) !== 0)
    throw new Error("Launcher integration failed");
  if (requests !== 1)
    throw new Error(`Expected one inference; got ${requests}`);
  console.log("PASS: launcher reused the external model without stopping it");
} finally {
  model.stop(true);
  await rm(profile, { recursive: true, force: true });
}

File: scripts/test-launcher.py
"""Exercise the real app on a private IBus bus with a deterministic local model."""
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
import json
import os
from pathlib import Path
import subprocess
import sys
import tempfile
import threading
import time

PROJECT = Path(__file__).resolve().parents[1]


def main():
    requests = []
    class Handler(BaseHTTPRequestHandler):
        def log_message(self, *_):
            pass

        def reply(self, value):
            data = json.dumps(value).encode()
            self.send_response(200)
            self.send_header("Content-Type", "application/json")
            self.send_header("Content-Length", str(len(data)))
            self.end_headers()
            self.wfile.write(data)

        def do_GET(self):
            if self.path == "/health":
                self.reply({"status": "ok"})
            elif self.path == "/v1/models":
                self.reply({"data": [{"id": "qwen3-1.7b"}]})
            else:
                self.send_error(404)

        def do_POST(self):
            if self.path != "/v1/chat/completions":
                self.send_error(404)
                return
            body = json.loads(self.rfile.read(int(self.headers["Content-Length"])))
            requests.append(1)
            line = next(line for line in body["messages"][1]["content"].splitlines() if '"程式"' in line)
            time.sleep(.1)
            self.reply({"choices": [{"message": {"content": json.dumps({"best": int(line.split(".")[0])})}}]})

    app = Path(os.environ.get("RIWEN_TEST_APP", PROJECT))
    root = Path(os.environ.get("RIWEN_DATA_DIR", app / ".cache"))
    root.mkdir(parents=True, exist_ok=True)
    model = ThreadingHTTPServer(("127.0.0.1", 0), Handler)
    thread = threading.Thread(target=model.serve_forever, daemon=True)
    thread.start()
    try:
        with tempfile.TemporaryDirectory(prefix="t-", dir=root) as profile:
            source = Path(os.environ.get("RIWEN_RIME_SOURCE") or Path(os.environ.get("XDG_CONFIG_HOME") or Path.home() / ".config") / "ibus/rime")
            env = dict(os.environ, RIWEN_MODEL_URL=f"http://127.0.0.1:{model.server_port}/v1/chat/completions",
                       RIWEN_PROFILE=profile, RIWEN_SESSION_DIR=f"{profile}/session",
                       RIWEN_SCHEMA="rime_frost_double_pinyin_flypy", RIWEN_RIME_SOURCE=str(source))
            subprocess.run(["dbus-run-session", "--", sys.executable, str(PROJECT / "tests/ibus_integration.py"), profile, "--launcher"],
                           cwd=PROJECT, env=env, timeout=90, check=True)
            assert requests == [1], f"Expected one inference; got {len(requests)}"
            assert thread.is_alive(), "Launcher stopped an external model"
            print("PASS: Python launcher reused the external model without stopping it")
    finally:
        model.shutdown()
        model.server_close()
        thread.join()


if __name__ == "__main__":
    main()
